### DISCUSSION TITLE
Is there a way to add a favicon?

### DISCUSSION THREAD

**Discussion by redtrillix** [rejetto/hfs#94] [Q&A] [answered]
I looked in the to-do and couldn't see anything about favicons so forgive me if I'm blind, but I was wondering if there is a way to have favicons, I've tried everything I can think of like having a 'favicon.ico' file in the root of both the program and the vfs and also tried the custom header with some html with no luck (from what I understand, it it used for the body so it makes sense it doesn't work). 

Not sure if it is a limitation of node.js or something, but I believe it would be a great addition to the customisation we already have with hfs.

**Comment by rejetto** [comment]
the favicon is in the < head> part, while the header feature inserts stuff in the < body> part, because it's where the visible stuff is meant to be.

i guess it's technically possible to put some javascript there, to manipulate the < head>, but probably not ideal.
Not a limitation of node for sure, i just didn't work on it.
I just have to decide what's the best way, what's the best tool to give the user that will allow him to set the favicon.
A dedicated "favicon" option may be easier/clearer, even tho not powerful (you only do the favicon).

**Reply by redtrillix** [reply]
I would say, seeing as a favicon is a simple thing not much configuration is needed and we probably dont need anything too powerful, especially as a first implamentation.

**Reply by redtrillix** [reply]
> i guess it's technically possible to put some javascript there, to manipulate the < head>, but probably not ideal.

For now I might see if theres a way to manipulate the header to get one working, just for the fun of it really.

**Reply by rejetto** [reply]
if you want to try, you add the < script> and use javascript to add something inside the < head>
but requires you to know javascript, or try adapting some code you find around.
LMK

**Reply by rejetto** [reply]
go chatgpt :D
<img width="909" alt="image" src="https://user-images.githubusercontent.com/1367199/216015628-1f4ab139-9484-4b8e-b2cc-36ba57d36595.png">

**Reply by redtrillix** [reply]
yeah chatgpt has been quite helpful for code

**Comment by redtrillix** [comment]
You could probably do something like what you've done with html files and have an option to show an image or icon as a favicon.
<br>
Like this (just an example): 
![image](https://user-images.githubusercontent.com/54786587/216015650-120199c1-7804-45cb-97c5-6434d191d58e.png)

**Comment by rejetto** [comment]
well, i wasn't considering that serving /favicon.ico may actually be enough.
add it to your VFS and hide it.
try it :)

**Reply by redtrillix** [reply]
unfortunately i couldn't get it to work, i think it would have to be built into the html head so your browser knows to look for it, cause i dont believe browsers look for a `/favicon.ico` by default.

**Reply by redtrillix** [reply]
this is it with it in the root and still no results:

![image](https://user-images.githubusercontent.com/54786587/216017324-effa0e65-6feb-4411-a83b-d74edb6f5254.png)
---
![image](https://user-images.githubusercontent.com/54786587/216017549-f58e4823-30fe-41ec-87f9-74e532b1122c.png)

**Reply by rejetto** [reply]
i can clearly see the browser trying to load /favicon.ico without any change to the html
<img width="410" alt="image" src="https://user-images.githubusercontent.com/1367199/216017474-095e4ccd-6bee-4f45-b32c-db59885896a1.png">

try hard-reloading (right-click on the reload button)

**Reply by rejetto** [reply]
i just tried, and it's working here
Put this file (virtually renamed)
<img width="390" alt="image" src="https://user-images.githubusercontent.com/1367199/216019774-2e3fa771-9666-4a04-bdc5-d81c1d5b1f46.png">

and got this on the browser's tab
<img width="252" alt="image" src="https://user-images.githubusercontent.com/1367199/216019707-74c69378-62e3-4cd4-ad68-badb3a0d8a21.png">

**Reply by redtrillix** [reply]
~~thats weird, might be my browser but thats good to know, tried hard refreshing and even used incognito~~

**Reply by rejetto** [reply]
mine is just latest chrome

**Reply by redtrillix** [reply]
> ~thats weird, might be my browser but thats good to know, tried hard refreshing and even used incognito~

I figured it out, it must be cloudflare cdn blocking because i can see it with lan ip

**Reply by rejetto** [reply]
if it's a cdn i guess it's just a cache (temporary) problem

**Reply by redtrillix** [reply]
nah i have bypass cache enabled

**Reply by redtrillix** [reply]
ive also had the favicon there for a long time

**Comment by rejetto** [comment, marked as answer]
ok, now we know adding the favicon to the vfs is (sort of) enough.
Still I have to decide what to do about it.
Letting user guess it (or google for it) is not ideal.

**Comment by rejetto** [comment]
while technically unnecessary, best user experience may still be to have a dedicated option in "configuration" page, i guess

**Reply by redtrillix** [reply]
yeah especially if they dont know where to put it, it might also allow it to be hidden from the root of the vfs with the other files/folders.

**Reply by rejetto** [reply]
well, of course once configured it would not be included in the list

**Reply by redtrillix** [reply]
~~it might just be cached, but it seems that the favicon will still load if its hidden~~.

Nope l lied, it was cached, if its marked as not downloadable, it doesnt show up.

![image](https://user-images.githubusercontent.com/54786587/216034972-8299e1eb-47f0-4084-9d6b-f94ca9759023.png)
---
but if you mark it as hidden from view but downloadable, it still works as a favicon:
![image](https://user-images.githubusercontent.com/54786587/216035852-13ff7b17-760d-4e8f-b8e8-1b184bcdb8e5.png)

**Comment by rejetto** [comment]
are you english mother tongue?
Would you let me know if you see english quirks in the GUI ?

**Reply by redtrillix** [reply]
is english my primary language?

**Reply by rejetto** [reply]
yeah

**Reply by redtrillix** [reply]
yes it is

i can do that for you
so far i haven't seen any quirks or issues but i will let you know if i do

**Comment by rejetto** [comment]
<img width="536" alt="image" src="https://user-images.githubusercontent.com/1367199/217112762-8b0c03d5-c608-4a07-8567-831d55bbceb0.png">

in next release

**Reply by HiCyprus** [reply]
![Screenshot 2025-01-23 010752](https://github.com/user-attachments/assets/492fddbb-6bcc-40f9-bb30-9fc1c6480387)
Seems that only JPG files can be added as Favicon here in the Options Pages. But... JPG is not good enough for icon files. Can we choose png or ico files here? Anyway, we could put the file in the root folder(the way which I haven't tried)

**Reply by rejetto** [reply]
@HiCyprus what makes you think that only jpg can be added?
i'm using a png, for instance

**Reply by HiCyprus** [reply]
![Screenshot 2025-01-24 222930](https://github.com/user-attachments/assets/8305bb5d-f75f-40bc-a2fe-614c1cc73bd6)
Thanks for the reply. I typed in the png file name directly and succeeded. But if I use the above UI, then only jpg files will be displayed to be chosen!

**Reply by rejetto** [reply]
right, i forgot to include png files in the list. Thanks for reporting. Will be fixed in next release
